app/common/metrics.py

```diff
@@ -9,3 +12,48 @@
     'build_state_duration_seconds',
     'Total amount of time a build spends in each build state',
     ['state'])
+
+serialized_build_time_seconds = Histogram(  # pylint: disable=no-value-for-parameter
+    'serialized_build_time_seconds',
+    'Total amount of time that would have been consumed by builds if all work was done serially')
+
+build_estimation_error = Histogram(  # pylint: disable=no-value-for-parameter
+    'build_estimation_error',
+    'For each build, the sum of the absolute value of the difference between the estimated and actual time for each atom to complete')
+
+
+class SlavesCollector(object):
+    """
+    Prometheus collector to collect the total number of alive/dead/idle slaves connected to the master.
+    collect() is called once each time prometheus scrapes the /metrics endpoint. This class ensures that
+    1. The list of slaves only gets iterated through once per scrape
+    2. A single slave is is not double counted in 2 states
+    """
+    def __init__(self, get_slaves: Callable[[], List['app.master.slave.Slave']]):
+        self._get_slaves = get_slaves
+
+    def collect(self):
+        active, idle, dead = 0, 0, 0
+        for slave in self._get_slaves():
+            if not slave.is_alive(use_cached=True):
```

Comment:
TJ, it sounds like you might be hinting at #287, which I still think we should do but probably not like this. If we set `use_cached=False` here I think every time Prometheus went to scrape metrics (every 15 seconds?) we'd synchronously send a request to every slave. I think this would not be the most efficient way to do heartbeats to slaves and we definitely wouldn't want to do it while we were handling an API request.

@ethomas2 I was assuming that `collect()` gets called when Prometheus scrapes the metrics endpoint, but is that correct?